Feature/refactor orb slam2 front end to accept arbitary data sources and data sinks

apattnaik0721013 opened the pull request:
Currently kitty data source , semantic data source is supported.
Following design is implemented.
l
![slam_application_process](https://user-images.githubusercontent.com/30471366/38409302-81e4caa8-399e-11e8-8d1d-c51619fd4629.png)

Commit:
add generic data source support

Commit:
add generic data source support

paul-michalik commented on ext/data_source.hpp:
```diff
@@ -0,0 +1,84 @@
+
+namespace ORB_SLAM2 {
+    namespace ext {
+
```
The types of data source should not be known here at all. Assume we want to assist source implementers. Then, look at [``boost::iterator_facade``](https://www.boost.org/doc/libs/1_66_0/libs/iterator/doc/iterator_facade.html) pattern (= CRTP). One possibility would be to create a simplified façade for users to define own iterators for own data sources:
```
template<class TDerived>
class data_source_iterator_facade: public boost::iterator_facade<
    TDerived,
    ext::tuple<ext::time_point_t, ...> // etc...
    boost::forward_traversal_tag
> {
// ...
};
```
Then the implementation of a new data source would consist of implementing a new iterator using the `data_source_iterator_facade<my_data_source_iterator>`. However the benefits from this would be marginal, I'd prefer to expose `iterator_facade` directly.

Second possibility would be to provide a façade for a data source, i.e. `data_source_facade<TDerived>` which would basically implement a [`SinglePassRange` concept](https://www.boost.org/doc/libs/1_66_0/libs/range/doc/html/range/concepts/single_pass_range.html). However I am not sure if implementers could benefit from this too much either. 

Either of the possibilities would make sense if we provided a concept check for user implementation but even that would have to be a specialization of [`SinglePassRangeConcept`](https://www.boost.org/doc/libs/1_66_0/libs/range/doc/html/range/concepts/concept_checking.html)

paul-michalik commented on ext/run.cpp:
```diff
@@ -0,0 +1,31 @@
+#include "ext/ds_kitty.hpp"
+#include "ext/ds_dashcam.hpp"
+#include "ext/run_tracking.hpp"
+
```
Imagine we have a set of data sources implemented following the pattern described above - i.e. each data source implements a `boost::iterator_facade<ext::ds_kitty_iterator, ...>`. Then the implementation of a program which uses specific data source looks like this: 
```
#include "ext/run_tracking.h"
#include "ext/ds_kitty.h"
#include "boost/ranges.hpp"

int main(argc, argv)
{
    ext::ds_kitty_args args{argc, argv};
    ext::run_tracking(boost::make_iterator_range(ext::ds_kitty_iterator{args}, ext::ds_kitty_iterator{}));
}
```

Alternatively, if a data source is modeled by a specific range:
```
#include "ext/run_tracking.h"
#include "ext/ds_kitty.h"
#include "boost/ranges.hpp"

int main(argc, argv)
{
    ext::ds_kitty_args args{argc, argv};
    ext::run_tracking(ext::ds_kitty{args});
}
```

So no need for case distinctions, just have a distinct runner for any ds type...

apattnaik0721013 replied:
@paul-michalik 
I have the following suggestion:
As we already made orb-slam2-ext has separate library,semantic_monocular.exe as separate application, cant we make  orb-slam2-ext to handle all user input so that from application point of view (semantic_monocular.exe),  the particular application should not know which data source it use, more specifically its implementation. so that we have control over is data source to change at latter timer, if any changes required.
which means from application point of view: the pseducode can be
#include "ext/run_tracking.h"

int main(argc, argv)
{
    ext::slam_helper slam(args,argv);  //it is wrapper class which will handle calling orb_slam2 object
                                                         //currently it is created as ext/slam_object
    slam.run_tracking();
}

1. I believe , though implementation is same for both cases, but with this we have same application which can run any type of application irrespective of input data type.
2. in code wise what you described , we can move to orb-slam2-ext library only. so in most of time we will change only orb-slam2-ext library only. (close for modification and open for extension)
Let me know if it make any valuable meaning.

paul-michalik commented:
I guess it’s better to deal with user input separately, specific to each data source. The proposed structure of the application should be as proposed in the review:

  1.  Data source: retrieves the data and provides a single pass range of sensor readings to the application. It deals with all data source specific user input
  2.  The SLAM object. Deals with system initialization, GUI interaction, controls the lifetime of the system and forwards the tracking calls to the system.
  3.  The tracking loop. Receives a data source and SLAM object. Pulls data from the data source and feeds them to the SLAM pipeline.
  4.  The runner. Instantiates and initializes the data source, SLAM object and starts the tracking loop. Deals with inter thread communication when necessary. There is one runner for each type of data source.
Please do not introduce names like slam_helper. The semantic_monocular will be replaced by data source specific runners which look as in the code snippet from the review. The implementation might get more complex for real time runners. Small and simple runners remove the necessity to deal with argument dispatching, factories and similar stuff…


Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________
From: ansuman pattnaik <notifications@github.com>
Sent: Tuesday, April 10, 2018 7:01:51 AM
To: paul-michalik/ORB_SLAM2
Cc: Paul Michalik; Mention
Subject: Re: [paul-michalik/ORB_SLAM2] Feature/refactor orb slam2 front end to accept arbitary data sources and data sinks (#68)


@apattnaik0721013 commented on this pull request.

________________________________

In ext/run.cpp<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23discussion_r180297634&data=02%7C01%7C%7C40d28122fbeb43778f0308d59ea02ddb%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636589333150090987&sdata=Umn8LTmh5WqiKkZOlkreVcVJcVek64aAdJR3vBLqp7M%3D&reserved=0>:

> @@ -0,0 +1,31 @@
+
+#include "ext/slam_api.hpp"
+#include "ext/ds_semantic.hpp"
+#include "ext/ds_kitty.hpp"
+#include "ext/ds_dashcam.hpp"
+#include "ext/run_tracking.hpp"
+


@paul-michalik<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik&data=02%7C01%7C%7C40d28122fbeb43778f0308d59ea02ddb%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636589333150090987&sdata=yNrhYtMnEcsHTOqzeMQ2URb7wHHdpMwqnGCXxznIM4M%3D&reserved=0>
I have the following suggestion:
As we already made orb-slam2-ext has separate library,semantic_monocular.exe as separate application, cant we make orb-slam2-ext to handle all user input so that from application point of view (semantic_monocular.exe), the particular application should not know which data source it use, more specifically its implementation. so that we have control over is data source to change at latter timer, if any changes required.
which means from application point of view: the pseducode can be
#include "ext/run_tracking.h"

int main(argc, argv)
{
ext::slam_helper slam(args,argv); //it is wrapper class which will handle calling orb_slam2 object
//currently it is created as ext/slam_object
slam.run_tracking();
}

  1.  I believe , though implementation is same for both cases, but with this we have same application which can run any type of application irrespective of input data type.
  2.  in code wise what you described , we can move to orb-slam2-ext library only. so in most of time we will change only orb-slam2-ext library only. (close for modification and open for extension)
Let me know if it make any valuable meaning.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23discussion_r180297634&data=02%7C01%7C%7C40d28122fbeb43778f0308d59ea02ddb%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636589333150090987&sdata=Umn8LTmh5WqiKkZOlkreVcVJcVek64aAdJR3vBLqp7M%3D&reserved=0>, or mute the thread<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKF8bt3i5IXtVq4NV66O-r7-d0HbRP6Wks5tnDy_gaJpZM4TJpfe&data=02%7C01%7C%7C40d28122fbeb43778f0308d59ea02ddb%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636589333150090987&sdata=6b7lff%2FlG11BUXV04L0WkXWYJniUFNxs%2FGbYgfBXEeE%3D&reserved=0>.

paul-michalik commented:
Version with `slam_object` created outside of `run_tracking`..
```
#include "ext/run_tracking.h"
#include "ext/slam_object.h"
#include "ext/ds_kitty.h"
#include "boost/ranges.hpp"

int main(argc, argv)
{
    ext::ds_kitty_args args{argc, argv};
    ext::slam_object slam{argc, argv};
    ext::run_tracking(slam, boost::make_iterator_range(ext::ds_kitty_iterator{args}, ext::ds_kitty_iterator{}));
}
```

apattnaik0721013 commented:
@paul-michalik 
I tried to implement your comments. Please review the changes.

paul-michalik commented on CMakeLists.txt:
```diff
@@ -53,6 +53,16 @@ add_library(orb-slam2-ext STATIC
     ext/app_monitor_api.h
     ext/app_monitor_api_impl.h
     ext/messages.h
+    ext/kitty_data_source/ds_kitty.hpp
```
Make it shorter: `ext/ds_kitty/...`. But do we really need a folder?

paul-michalik commented on CMakeLists.txt:
```diff
@@ -141,14 +151,24 @@ if(BUILD_ALL_EXAMPLES)
 elseif(BUILD_EXAMPLES)
-    add_executable(semantic_monocular Examples/Monocular/semantic_monocular.cc Examples/Monocular/semantic_monocular.hpp)
-    target_link_libraries(semantic_monocular 
+    add_executable(run_kitty_monocular Examples/Monocular/run_kitty_monocular.cc)
```
Please use the name of the data source in the target and source file names: `run_ds_kitti.cpp`, `run_ds_video.cpp`, ...

paul-michalik commented on Examples/Monocular/run_dashcam_monocular.cc:
```diff
@@ -0,0 +1,66 @@
+
```
1. Factor the common parts of each runner implementation into a new module.
2. Data source implementation must export an API to create itself. A function template to create a data source:
```
// TDataSource is a trait class which describes the data source and is provided by a data source implementation
auto make_data_source<TDataSource>(argc, argv)
{
    return boost::range::make_iterator_range<
        typename TDataSource::iterator_t, ... // etc, you get the point
    >(argc, argv);
}
```
3. Rewrite runners according to this API

apattnaik0721013 replied:
small doubt on this.
1. do you want to make a common iterator template class for all data source?
2. from data source, do you want to implement begin() and end() , so that it can be used as stl container (example: TDataSource::iterator = TDataSourceObj.begin())
3. in run_tracking.hpp, we are going to add make_data_sourcec() which will return range for specific data source.

paul-michalik reviewed (changes requested):
Please rename `*. cc` to `*. cpp` and stay consistent. The same for `*.h` or `*. hpp`. Please use either of them but  not both.

paul-michalik commented on Examples/Monocular/testJson.json:
```diff
@@ -1,156 +1,274 @@
-{
-  "000000": {
-    "traffic_signs": [
+[
```
Please move this into a new test folder. I suggest `tests/tests_ds_semantic.json` and all other unit test shall go there and shall be named following a unified schema.

paul-michalik commented on ext/ds_dashcam.hpp:
```diff
@@ -0,0 +1,188 @@
+namespace fs = boost::filesystem;
+using namespace std;
+using boost::property_tree::ptree;
+
```
Are we reading the csv file here? I'd prefer to red the mpeg stream directly!

paul-michalik commented on ext/run.cpp:
```diff
@@ -0,0 +1,31 @@
+#include "ext/ds_dashcam.hpp"
+#include "ext/run_tracking.hpp"
+
+namespace ORB_SLAM2 {
```
Why is this still there? Please remove...

paul-michalik commented on scripts/windows/run_semantic_monocular.bat:
```diff
@@ -16,11 +16,11 @@ if not exist "%OrbSlam2_ProjectDir%\Vocabulary\ORBvoc.bin" (
 )
 
-set "OrbSlam2_App=%OrbSlam2_ProductsDir%\Release\semantic_monocular.exe"
+set "OrbSlam2_App=%OrbSlam2_ProductsDir%\Release\run_semantic_monocular.exe"
```
This file needs to be completely removed too, or am I misunderstanding something?

paul-michalik reviewed (changes requested):
The runners are obviously still far away from the ideal we have set up a while ago. There's too much code in them, most of it repeated over all nrunners.

paul-michalik commented:
The ideal I mentioned in the review comment is this (grabbing "kitti" data source as an example)... Alternative is the iterator range based version. Let us work towads this result step by step!

```
#include "ext/run_tracking.h"
#include "ext/slam_object.h"
#include "ext/ds_kitty.h"

int main(int argc, char** argv)
{
    ext::ds_kitty_args args{argc, argv};
    ext::slam_object slam{argc, argv};
    ext::run_tracking(slam, ext::ds_kitty{args});
}
```

apattnaik0721013 commented:
@paul-michalik 
_following changes are made:_
1. rebase master
2. use consistent header and source file extension(.cc and .h)
3. remove run_semantic_monocular.bat
4. moved Examples/Monocular/testJson.json to tests/testJson.json
5. remove ext/run.cpp
6. change run_ds_kitty.cpp ,run_ds_dashcam.cc and run_ds_semantic.cc to the following:

```
//  in run_ds_kitty.cpp:
#include "ext/run_tracking.h"
#include "ext/slam_object.h"
#include "ext/ds_kitty.h"

int main(int argc, char** argv)
{
    ext::ds_kitty_args args{argc, argv};
    ext::slam_object slam{argc, argv};
    ext::run_tracking(slam, ext::ds_kitty_iter{args});
}
```
**please note:**
- though we don't need separate iterator class ds_kitty_iter (ds_kitty_iter.h) for class ds_kitty as we can make class ds_kitty iterator class, but with class ds_kitty_iter, it makes more simple and clean.

**to be done:**
1. provide iterator range based version.
2. update ds_dashcam.h for reading mpeg stream directly

@paul-michalik let me know your input on this.

paul-michalik commented on ext/ds_kitty.h:
```diff
@@ -0,0 +1,159 @@
+					}
+					gpsfile.close();
+					_gps_index++;
+					//slow down driving speed for avoiding tracking lost
```
What is this for?

paul-michalik replied:
This isn't good at all. Has Poorva run into this when testing against Kitti datasets for the first time?

paul-michalik replied:
We need to investigate this. @apattnaik0721013's theory is that the mapping thread takes too much time and in consequence the next frame from the tracking differs too much from the last seen frame (=loss of coherence). 
1. Why does the mapping thread take much longer than with the other data sets? Especially the Garching data appears to be more complex than the Kitti data...
2. I propose to enlarge the size of the queue which is used for communication between tracking and mapping threads. This should help amend the loss of tracking...
3. Are the configuration data used for the feature detector the same as for other data sets? Have you tried we lower the number of detected features?

apattnaik0721013 replied:
q. why tracking lost occurs?
Ans. unrelated neighbor images are encountered.
Description ->
- neighbor images can be unrelated if the matching of features(key-points) are less among neighbor images.
- In monocular orb-slam2, map-points are considered if there is at-least 2 observers which means at least 3 neighbor key-frames. 
- each key-frame is constructed after certain frames is passed. so there may be 2 or more frames(images) should be passed for each key-frame construction.
- so in assumption, features(key-points) of one images should have more matches with its 4th image.
causes of unrelated neighbor images ->
1. vehicle is moving very fast which may create more difference in terms of matching of features(key-points) among neighbor images or frames.
2. features(key-points) extraction is less in images.
3. creation of key-frames becomes less as local mapping takes huge time to process each key-frame.
- local mapping perform following task when it encounters a key-frame in the queue->
-- MapPointCulling()                 - remove redundant local map-points
-- CreateNewMapPoints()              - create map-point from current key-frame
-- SearchInNeighbors();              - replace duplicate map-points
-- Optimizer::LocalBundleAdjustment  - optimizations of map-points
-- KeyFrameCulling()                 - remove redundant redundant local Key-frames

q1.Why does the mapping thread take much longer than with the other data sets? Especially the Garching data appears to be more complex than the Kitti data...
Ans.  
- There cant be any particular answer as it depends on matching of features(key-points) among images . 
- But my assumption is that in some kitty images, either vehicle is move very fast and kitty fps is 10 per sec which my casuse more difference among images in terms of features(key-points).

q2. I propose to enlarge the size of the queue which is used for communication between tracking and mapping threads. This should help amend the loss of tracking...
Ans
- yes we can enlarge the size of the queue to avoid tracking lost. Though we can use sleep() in tracking which also increase more keyframe. but use queue will give more control to select particular frame to become key-frame.
- in orb-slam2 (raulmur version) , sleep() is used to avoid tracking lost.
- the key-frame is filtered by orb-slam2 in order to view the result in sync with the current input though run tracking , local mapping and viewer thread are run asynchronously. 
- Though we can make more key-frames or all the key-frames from frames and process them all in local mapping, but this will make orb-slam2 to run tracking and local mapping thread synchronously and viewer thread more out-of-sync with the current input. so the lag will seen between current input frame and  map-viewer.
- so it is important to select key-frames more carefully to avoid performance issue.

q3.Are the configuration data used for the feature detector the same as for other data sets? Have you tried we lower the number of detected features?
Ans
- i tried to run kitty data set with lower features and higher features , but result is same.

paul-michalik commented:
The runners look very good now. Now looking into the implementation of the data sources strange things are coming up...: Which of the data sources are really functional? How do you test this? More ds specific questions in the comments...

paul-michalik commented:
Please look up the JIRA task where I describe how to extract the GPS which are encoded as subtitles from the transport stream. Not sure if cv::VideoCapture supports this. Maybe via properties, similar as we did for the time stamp? Otherwise use ffmpeg? Please investigate!

Sent from my Windows 10 phone

________________________________
From: ansuman pattnaik <notifications@github.com>
Sent: Thursday, May 3, 2018 10:22:50 AM
To: paul-michalik/ORB_SLAM2
Cc: Paul Michalik; Mention
Subject: Re: [paul-michalik/ORB_SLAM2] Feature/refactor orb slam2 front end to accept arbitary data sources and data sinks (#68)


@apattnaik0721013 commented on this pull request.

________________________________

In ext/ds_dashcam.hpp<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23discussion_r185724801&data=02%7C01%7C%7C49a4bc185f10423e21d208d5b0cf101d%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636609325725877965&sdata=gcMbbiljYJhuZharQFwp6u6vqvfiuSBRPrdsMNFKZfc%3D&reserved=0>:

> +#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/foreach.hpp>
+#include <iostream>
+#include <boost/filesystem.hpp>
+#include<iterator>
+#include <boost/lexical_cast.hpp>
+#include<opencv2/core/core.hpp>
+#include "opencv2/opencv.hpp"
+#include <boost/tokenizer.hpp>
+#include <boost/lexical_cast.hpp>
+
+namespace fs = boost::filesystem;
+using namespace std;
+using boost::property_tree::ptree;
+


@paul-michalik<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik&data=02%7C01%7C%7C49a4bc185f10423e21d208d5b0cf101d%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636609325725877965&sdata=RNCsSrNoceBFnuVdry4JwWKFhMvZZ%2BYTz8nAP5BBumA%3D&reserved=0>

  *   I am reading mpeg stream. and also csv file.
  *   mpeg steam for extracting images
  *   csv file for extracting gps information.

let me know if we any other ways are there to extract gps.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23discussion_r185724801&data=02%7C01%7C%7C49a4bc185f10423e21d208d5b0cf101d%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636609325725877965&sdata=gcMbbiljYJhuZharQFwp6u6vqvfiuSBRPrdsMNFKZfc%3D&reserved=0>, or mute the thread<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKF8bnMbFJNHlulVwegntIfUXCgq1u0lks5tur5agaJpZM4TJpfe&data=02%7C01%7C%7C49a4bc185f10423e21d208d5b0cf101d%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636609325725877965&sdata=5qKWiHm0y1s6YvVEszwZjFgYgpBksOP7NuSPg4eCdoE%3D&reserved=0>.

paul-michalik commented:
Please add an (intermediate) Readme in the `ext` with brief description how to use the runners, which CL arguments have to set what is their semantics.

paul-michalik commented on Examples/Monocular/testJson.json:
```diff
@@ -1,156 +0,0 @@
-{
```
Don't use CamelCase in file names. Prefer `some_meaningful_name.json`

paul-michalik replied:
If you have test data, please add it to a new folder called `ORB_SLAM2/tests`. If you feel confident and want to make me happy add also unit tests which verify the correctnes of the current implementation if the respective data sources.

apattnaik0721013 commented:
@paul-michalik 

1. add readme to use the application
2. add script for the extended application.

paul-michalik committed:
Fixed most obvious typos and errors in Readme

paul-michalik commented on ext/Readme.md:
```diff
@@ -0,0 +1,23 @@
+- dashcam: The MPEG transport stream generated by Ausdom A261 DashCam. [Format description and analysis](https://jira.harman.com/jira/browse/INTHAD-183?filter=53275&jql=project%20%3D%20%22Highly%20Automated%20Driving%22%20AND%20%20labels%20in%20(Macs-SLAM-ORB2)%20AND%20description%20~%20Ausdom%20ORDER%20BY%20created%20DESC)
+
+## Bootstrap and build
+
```
Please add the usual description about using the project API

paul-michalik commented on ext/Readme.md:
```diff
@@ -0,0 +1,23 @@
+
+### Run kitti
+* `run_ds_kitti.bat <path-to-image-folder> <path-to-oxts-folder> <path-to-settings-file>`
+* download Kitti dataset with GPS support http://kitti.is.tue.mpg.de/kitti/raw_data/2011_09_26_drive_0009/2011_09_26_drive_0009_sync.zip and unzip into a folder `kitti_data_dir`. To run the data set `00` invoke script as follows: `run_ds_kitti.bat kitti_data_dir\image_00\data kitti_data_dir\oxts\data`
```
How can this work? `image_00\data` contains images from one specific track. How is `oxts\data` related to `image_00`?

paul-michalik replied:
You need to document the connection between the input sequence and the raw data which contains the "oxts" readings. Please consult with @PTavse. Obviously not each input sequence will be compatible with arbitrary raw data folder!

apattnaik0721013 commented:
@paul-michalik 
following changes are done:
in run_tracking.h->

- waiting tracking time is calculated:

```
std::chrono::steady_clock::time_point start_time = std::chrono::steady_clock::now();
sys_.track(item);
std::chrono::steady_clock::time_point stop_time = std::chrono::steady_clock::now();
double actual_tracking_time = std::chrono::duration_cast<std::chrono::duration<double> >(stop_time - start_time).count();
std::this_thread::sleep_for(std::chrono::microseconds(wait_time));
```
- waiting tracking time is required in order to avoid tracking lost.
- this condition is already handled in orginal orb-slam2 of raulmur.
- reason of tracking lost is already mentioned in previous comment. 
- so data-source classes (ds_semantic.h,ds_dashcam.h,ds_kitty.h) is changed to calculate timestamp between two frames.

kitty data set -: 
There two sets of kitty datasets 
1. standard kitt datasets which contain images set , timestamp , and its ground truth. 
(http://www.cvlibs.net/datasets/kitti/eval_odometry.php)
 2. raw kitt datasets which contain images set , timestamp and gps data with ground truth.
(http://www.cvlibs.net/datasets/kitti/raw_data.php)

- both the data set are related. which means gps data can be taken from raw kitt datasets and and test with corresponding dataset with standard kitt datasets.
- mapping between raw datasets and kitti dataset can be found (https://github.com/alexkreimer/odometry/blob/master/devkit/readme.txt)
- i have tested with image and timestamp (data_odometry_gray\dataset\sequences\04) and gps data(2011_09_30_drive_0016_sync\2011_09_30\2011_09_30_drive_0016_sync\oxts\data)
by downloading image sequence and timestamp from  standard kitty dataset (http://www.cvlibs.net/download.php?file=data_odometry_gray.zip)
and raw dataset for gps from (http://kitti.is.tue.mpg.de/kitti/raw_data/2011_09_30_drive_0016/2011_09_30_drive_0016_sync.zip )

apattnaik0721013 commented:
@paul-michalik 

- removed ds_kitty_iter.h , ds_semantic_iter.h , ds_dashcam_iter.h.
- iterator functionalities are handled in ds_kitty.h , ds_semantic.h , ds_dashcam.h

Commit:
fix run script

Commit:
fix run script

Commit:
Feature/provide gps information from ausdom a261 dash cam to orb slam2 (#79)

* add dashcam gps support

* fix gps synchronization

* use boost:split to parse gps string

Commit:
Feature/provide gps information from ausdom a261 dash cam to orb slam2 (#79)

paul-michalik committed:
(1) CppProperties was wrong and consistently broke IntelliSense (2) Add run_ds_dashcam.exe to launch.vs.json

paul-michalik committed:
Fix CMake generation issues

Commit:
add gps parser in util library

apattnaik0721013 commented:
@paul-michalik 
- move  ffmpeg utility from ds_dashcam.h to utils library
- move gps parser from ds_dashcam.h to utils library
- copy experiments\cpp\geoconv2\utils\slam_utils library to util\geo_convertor 
- remodel ds_dashcam.h and ds_kitty.h to use utils library.
- run_tracking.h is customized to add some default wait time. currently it is only needed for kitty dataset.

status:
- we can run kitty dataset with gps support through run_ds_kitty.bat
- we can dashcam videos embedded with gps through run_ds_dashcam.bat
- we can run semantic data with images through run_ds_semantic.bat

Commit:
add header files

paul-michalik committed:
(1) Update project API files for Windows to current state (2) Add missing artifacts

paul-michalik committed:
Add .vendor path to .gitignore

Commit:
fix default wait time

paul-michalik reviewed (commented):
How is the wait time determined? It should be retrieved from the configuration as fps. The runner should assure that tracking thread does not exceed the required fps.

apattnaik0721013 commented:
@paul-michalik 
- ``wait_time of current frame = (timestamp of next frame - timestamp of current frame) * default_time``
- the ``default_time`` of kitti dataset = 1e6. this value is set by raumlur.
- for other dataset, ``default_time`` is 1.

paul-michalik commented:
Ansuman, this was a requirement not a question. We need to make sure that the tracking thread does not exceed the required  frame rate. Please hold on with pushing commits to this branch I need to fix something fundamental here.

Sent from my Windows 10 phone

________________________________
From: ansuman pattnaik <notifications@github.com>
Sent: Thursday, July 26, 2018 10:39:53 AM
To: paul-michalik/ORB_SLAM2
Cc: Paul Michalik; Mention
Subject: Re: [paul-michalik/ORB_SLAM2] Feature/refactor orb slam2 front end to accept arbitary data sources and data sinks (#68)


@paul-michalik<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik&data=02%7C01%7C%7Cd7028481753b485bfeef08d5f2d35d32%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636681911966818709&sdata=q%2BckgdJAwUALiNvSIprxZOxMJ97vyrZ1KQHLkW5EeYo%3D&reserved=0>

  *   wait_time of current frame = (timestamp of next frame - timestamp of current frame) * default_time
  *   the default_time of kitti dataset = 1e6. this value is set by raumlur.
  *   for other dataset, default_time is 1.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23issuecomment-408022783&data=02%7C01%7C%7Cd7028481753b485bfeef08d5f2d35d32%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636681911966818709&sdata=D1Bb3rAHpdz5zXvBjebaIjZ0ljvrN%2FTx5Oi16BFzISY%3D&reserved=0>, or mute the thread<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKF8bt1Yv6s1G2wOupEWVYep5g5T3VaWks5uKYBYgaJpZM4TJpfe&data=02%7C01%7C%7Cd7028481753b485bfeef08d5f2d35d32%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636681911966818709&sdata=G%2Fwfd%2B4rDuvaIKvyKK6vGBM6eAdzBRCfogKtVTVBqpQ%3D&reserved=0>.

paul-michalik commented on ext/ds_semantic.h:
```diff
@@ -0,0 +1,299 @@
+
+        class ds_semantic : public boost::iterator_facade<
+                            ds_semantic,
+                            std::tuple<time_point_t, image_t, tsr_info_opt_t, pos_info_opt_t>,
```
Make this a `const std::tuple<...>`. Then, the strange casts aren't necessary. Please apply to every data source.

paul-michalik commented on ext/ds_semantic.h:
```diff
@@ -0,0 +1,299 @@
+            }
+
+            //iterator
+            ds_semantic begin()
```
Remove! This isn't a member for a iterator! Apply to all data sources. Instead introduce a `make` helper function in file shared by all data sources like so:
```
template<class TDataSourceItr, typename... TArgs>
auto make_data_source(TArgs&&... args_ )
{
    return boost::make_iterator_range(TDataSourceItr{std::forward<TArgs>(args_)...}, TDataSourceItr{});
}
```

The usage at client site: 

```
auto ds = ext::make_data_source<specific_data_source>(a, b, c, d);
```

paul-michalik commented on ext/ds_semantic.h:
```diff
@@ -0,0 +1,299 @@
+#pragma once
```
Rename this thing to something meaningful. `ds_semantic` is an idiotic name.

paul-michalik commented on ext/ds_semantic.h:
```diff
@@ -0,0 +1,299 @@
+            ds_semantic end() { return ds_semantic(); }
+            bool equal(const ds_semantic& other) const { return _image_index == other._image_index; }
+            void increment() { get_next_item(); }
+            std::tuple<time_point_t, image_t, tsr_info_opt_t, pos_info_opt_t> & dereference() const { return (std::tuple<time_point_t, image_t, tsr_info_opt_t, pos_info_opt_t> & const)_item; }
```
Apply to all data source iterators:

```
auto dereference() const 
{
    return _item;
}
```

paul-michalik commented on ext/run_tracking.h:
```diff
@@ -0,0 +1,59 @@
+
+        //The Range variant:
+        template<class TDataSource>
+        void run_tracking(slam_object& sys_, TDataSource data_source_,double default_wait_time= 1)
```
We need to pass a maximum required fps rate here which is passed based on the information form the environment. In all implementation it should be read from the yaml files.

```
/// Run tracking for given data source.
/// <param name="system_">What is this?</param>  
/// <param name="data source_">What is this?</param>  
/// <param name="max_fps_">Maximum allowed frame rate in frames per second. The tracking thread will never exceed this frame rate.</param>  
template<class TDataSource> 
void run_tracking(slam_object& system_, TDataSource data_source_, int max_fps_)
{
    // ...
}
```

Count the

paul-michalik reviewed (changes requested):
Please apply the changes to *all* data sources!

Commit:
add datasource range

apattnaik0721013 commented:
@paul-michalik 
- sorry for the delay. i was resolving an crash issue in data source.
- ds_semantic is changes to ds_tsr.
- need to use copy constructor in ds_tsr and ds_kitti, as these have std::fstream object which is not copyable.
- added make_data_source() which will give data_source range object.

paul-michalik committed:
Upgrade bootstrapping, helper and CMake scripts

paul-michalik committed:
Add g2o from head

Commit:
rename ds_tsr to ds_trafficsign

paul-michalik committed:
Renaming and fixing examples

paul-michalik commented:
@apattnaik0721013 
1. Please check the last commit. The acceptance test is `run_ds_tsr.exe` or ``run_ds_tsr.bat` needs to run with the `Garching_LoopClosure-5-images` dataset. It is not working right now.
2. Please ensure the constructors of each data source takes the respective `argument_type` instead of "raw" `(argc, argv)`

apattnaik0721013 commented:
@paul-michalik 
- i have received graching_loopcloser_5 dataset
- i guess the json format is not correct.  
- the actual format is as follows:
```{"image":"000000","time_point":"000000"}```
but Garching_LoopClosure-5-features.json contains following format-
```
  "000000": {
    "traffic_signs": [
      {
        "class_id": 35,
        "confidence": 0.9312705993652344,
        "rectangle": [
          0.32327091693878174,
          0.7908697128295898,
          0.3846966624259949,
          0.8184937238693237
        ]
      }
    ]
  }
```
- let me know how to process on this.

apattnaik0721013 commented:
```mistakenly clicked on `close and comment```

paul-michalik commented:
Oh, you’re right we’ve changed the format. Please generate the datasets with new format, commit them in the data folder. Please also commit the input video with git-lfs!

Sent from my Windows 10 device

________________________________
From: ansuman pattnaik <notifications@github.com>
Sent: Tuesday, July 31, 2018 4:52:41 PM
To: paul-michalik/ORB_SLAM2
Cc: Paul Michalik; Mention
Subject: Re: [paul-michalik/ORB_SLAM2] Feature/refactor orb slam2 front end to accept arbitary data sources and data sinks (#68)


@paul-michalik<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik&data=02%7C01%7C%7C3c2f193116324f412e3908d5f6f5457e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636686455640595821&sdata=sLf3s1%2BJmFcp9EDfxOJYTcHuLp9nFiZJZlRPy1c0by8%3D&reserved=0>

  *   i have received graching_loopcloser_5 dataset
  *   i guess the json format is not correct.
  *   the actual format is as follows:
{"image":"000000","time_point":"000000"}
but Garching_LoopClosure-5-features.json contains following format-

  "000000": {
    "traffic_signs": [
      {
        "class_id": 35,
        "confidence": 0.9312705993652344,
        "rectangle": [
          0.32327091693878174,
          0.7908697128295898,
          0.3846966624259949,
          0.8184937238693237
        ]
      }
    ]
  }


  *   let me know how to process on this.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23issuecomment-409248902&data=02%7C01%7C%7C3c2f193116324f412e3908d5f6f5457e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636686455640595821&sdata=vcPfhWnzNMbKTRfpF4yneYk5R97kBZENajn%2F2X1MXqE%3D&reserved=0>, or mute the thread<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKF8bh9Dby5jpq0AZUI2vBPM2c_W3yl1ks5uMG85gaJpZM4TJpfe&data=02%7C01%7C%7C3c2f193116324f412e3908d5f6f5457e%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636686455640595821&sdata=jZbqilX3pJgyNZridy4lnfEwcxVvOQ2hhnVanetaQoA%3D&reserved=0>.

paul-michalik commented:
Basically running the acceptance tests is the script run_ds_tsr.bat with default arguments succeeds. Please change the defaults to use video instead of image folder.

Sent from my Windows 10 device

________________________________
From: ansuman pattnaik <notifications@github.com>
Sent: Tuesday, July 31, 2018 4:55:19 PM
To: paul-michalik/ORB_SLAM2
Cc: Paul Michalik; Mention
Subject: Re: [paul-michalik/ORB_SLAM2] Feature/refactor orb slam2 front end to accept arbitary data sources and data sinks (#68)


mistakenly clicked on `close and comment

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23issuecomment-409249811&data=02%7C01%7C%7Cc4d4f0a1bb8d422113aa08d5f6f5a2bb%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636686457212795986&sdata=Yimdq865aKP3u%2BnufBjBy2eBy%2FWuixvz3Op9YDtMgaI%3D&reserved=0>, or mute the thread<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKF8biIaH3VizvTvqBmZA-FvvJ9KfTraks5uMG_WgaJpZM4TJpfe&data=02%7C01%7C%7Cc4d4f0a1bb8d422113aa08d5f6f5a2bb%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636686457212795986&sdata=LD6OL3Wm1XOqv9%2BVkMqNBYHvx7siv40mvohoRN8PgWw%3D&reserved=0>.

Commit:
use boost::program_options in data source

paul-michalik commented:
No, please pull this back! You overwrote the renames "traficsign" -> "tsr"...! Just remove this whole commit and start over!

Commit:
rename ds_trafficsign to ds_tsr

apattnaik0721013 commented:
@paul-michalik 
[Status]
- renames ds_trafficsign to ds_tsr
- regarding unit testing, we already have created this issue.
- so we can address this in separate branch
- currently ds_tsr.bat takes only images along with txt file

[Need feedback]
- do you want to change ds_tsr to handle both image and video input where default is video input.
- do i create test cases by using the graching_loopcloser_5 dataset in this branch

paul-michalik commented:
*   Just remove your commit and then add what you wanted to add without overwriting my changes
  *   What does ds_tsr support now? Video and images should be supported.
  *   Don’t create testcases now. It just has to work as described with a data set added using the given name. Please add that to data folder wiyh git-lfs for video and images.

Sent from my Windows 10 device

________________________________
From: ansuman pattnaik <notifications@github.com>
Sent: Wednesday, August 1, 2018 10:59:34 AM
To: paul-michalik/ORB_SLAM2
Cc: Paul Michalik; Mention
Subject: Re: [paul-michalik/ORB_SLAM2] Feature/refactor orb slam2 front end to accept arbitary data sources and data sinks (#68)


@paul-michalik<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik&data=02%7C01%7C%7C8873d36162484415400408d5f78d1ab9%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636687107758136618&sdata=qlB9neFOsMkLxV2kluYrqTEhOvjHPOQIG%2BKnctvA3Hc%3D&reserved=0>
[Status]

  *   renames ds_trafficsign to ds_tsr
  *   regarding unit testing, we already have created this issue.
  *   so we can address this in separate branch
  *   currently ds_tsr.bat takes only images along with txt file

[Need feedback]

  *   do you want to change ds_tsr to handle both image and video input where default is video input.
  *   do i create test cases by using the graching_loopcloser_5 dataset in this branch

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23issuecomment-409503919&data=02%7C01%7C%7C8873d36162484415400408d5f78d1ab9%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636687107758136618&sdata=VGYGhi0BRfXzW9hehCT4RwdXbogb6EAuJFtD4Ii7ucU%3D&reserved=0>, or mute the thread<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKF8bpn85shdoQIbuN6pwhAykDsX8Z1Kks5uMW32gaJpZM4TJpfe&data=02%7C01%7C%7C8873d36162484415400408d5f78d1ab9%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636687107758136618&sdata=WfbcSrRtbcJ%2FN3cHBbuoVLcphk7Y2gXcElHjK975bp4%3D&reserved=0>.

paul-michalik commented:
@apattnaik0721013 

Adding boost-parameters is a good thing, but a CLI should be **designed**. In your implementation there's a lot of typos and no recognizable system. The idea of the `arguments` types is to have a semantic type which represents the arguments of a specific data source and provides an **semantic** type safe API to retrieve the parameters independently of how they were named on the command line. I am going to do this for `ds_tsr_args`  and `ds_tsr`. Please do this for all other data sources after I commit the changes!

paul-michalik commented:
@apattnaik0721013  

**Stop adding `using namespace` directives to global namespace!**

apattnaik0721013 commented:
@paul-michalik 
sorry paul , i pushed my changes before seeing your latest comments.

paul-michalik commented:
@apattnaik0721013 

1. I am also going to change the code formatting. Please stick with using this formatting style! 
2. Don't use `BOOST_FOREACH`. There is range-base for loop!

paul-michalik committed:
(1) Add precompiled header support for run_ds_tsr (2) rename wait to throttle (2) Extracted semantic type for arguments to ds_tsr_args (3) Fixed typos

paul-michalik committed:
(1) Remove duplicated code in run_tracking (2) Rename throttle to wait_for with simpler signature

apattnaik0721013 commented:
@paul-michalik 
- i will remove BOOST_FOREACH and using namespace.
- as you mentioned , you are going to change `ds_tsr args` and `ds_tsr`, so after this i will address this changes.

paul-michalik commented:
No! I told you to wait until I push the changes to demonstrate how things should be done for ds_tsr. Then, please, go on with other data sources in the same style. You can branch off and continue there but I'll remove al the other changes from this branch. Also: From now on it is zero warnings policy and no feature without a test case. I’ll add cotired configuration for Boost Test to the project and change the build config accordingly.

Sent from my Windows 10 device

________________________________
From: ansuman pattnaik <notifications@github.com>
Sent: Friday, August 3, 2018 9:42:32 AM
To: paul-michalik/ORB_SLAM2
Cc: Paul Michalik; Mention
Subject: Re: [paul-michalik/ORB_SLAM2] Feature/refactor orb slam2 front end to accept arbitary data sources and data sinks (#68)


@paul-michalik<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik&data=02%7C01%7C%7C9560fd04fa0345a1e95d08d5f914acd1%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636688789541846414&sdata=PS%2BxAQPTU%2BZr0QVSKZv6B8lWLfliK5d408onC1itwq4%3D&reserved=0>

  *   i will remove BOOST_FOREACH and using namespace.
  *   as you mentioned , you are going to change ds_tsr args and ds_tsr, so after this i will address this changes.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23issuecomment-410172895&data=02%7C01%7C%7C9560fd04fa0345a1e95d08d5f914acd1%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636688789541846414&sdata=WbMXtFOS%2BqfK8FbmVg4TH91n5O%2BjnehK1eirKzNTnsM%3D&reserved=0>, or mute the thread<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKF8bmIwOhEuS_E757IOBkvXLdlkgEHoks5uM_7ogaJpZM4TJpfe&data=02%7C01%7C%7C9560fd04fa0345a1e95d08d5f914acd1%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636688789541846414&sdata=tYJGPiFlAJhckLOKE%2FZdcqhH%2Fxr8DjZcQjbRt9nN4C8%3D&reserved=0>.

paul-michalik committed:
(1) Move data sources to a new sub folder (2) Add ext_tests.boost (3) Set up cotire build

apattnaik0721013 commented:
@paul-michalik 
- sorry for this trouble.
- but i didnt update any code changes after your previous comments
> @apattnaik0721013
> I am also going to change the code formatting. Please stick with using this formatting style!
> Don't use BOOST_FOREACH. There is range-base for loop!
- so the last commit is https://github.com/paul-michalik/ORB_SLAM2/pull/68/commits/2a436fe3e41328ed130ef448734a527506dcdec8.
- but i added only the comments about the this commit
- just for information, this commit has following changes:
  - renames ds_trafficsign to ds_tsr
  - added graching data set
  - ad new iterator ds_image which will handle both image and video input in util
  - ds_kitti, ds_tsr, ds_dashcam use ds_image to handle both image and video input and provide slam input.
  - change launch.vs.json for launch run_dashcam and run_tsr.
  - run_ds_dashcam.bat and run_tsr.bat will launch with default graching dataset if not parameter is given
- let me know how to resolve this now.

paul-michalik commented:
This I fine, leave it be. I'll take care of it. I’m gonna do this:

  1.  Create a new branch from the tip of your current work. Rename it like th original branch plus ”-temporary” token
  2.  Reset the original branch to 6c202eb and push it
  3.  Add my changes to the original branch


Then, and only then you can start working on the discussed adaptations. You can switch to “-temporary” branch after it's ready.


Sent from my Windows 10 device

________________________________
From: ansuman pattnaik <notifications@github.com>
Sent: Wednesday, August 8, 2018 10:01:48 AM
To: paul-michalik/ORB_SLAM2
Cc: Paul Michalik; Mention
Subject: Re: [paul-michalik/ORB_SLAM2] Feature/refactor orb slam2 front end to accept arbitary data sources and data sinks (#68)


@paul-michalik<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik&data=02%7C01%7C%7C600db7eccf4f4b1c6ca908d5fd05319c%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636693121099696922&sdata=WkjvxQq940nda1D4zr41ZcqkQ9DYkAJ1PDSVN74uVWA%3D&reserved=0>

  *   sorry for this trouble.
  *   but i didnt update any code changes after your previous comments

@apattnaik0721013<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapattnaik0721013&data=02%7C01%7C%7C600db7eccf4f4b1c6ca908d5fd05319c%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636693121099696922&sdata=aDgwBUaN3IJN5VXR56sTGY6iN4tN3DlKSQ7HxFzilI8%3D&reserved=0>
I am also going to change the code formatting. Please stick with using this formatting style!
Don't use BOOST_FOREACH. There is range-base for loop!

  *   so the last commit is 2a436fe<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fcommit%2F2a436fe3e41328ed130ef448734a527506dcdec8&data=02%7C01%7C%7C600db7eccf4f4b1c6ca908d5fd05319c%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636693121099696922&sdata=BODnSlyQ72q2APuqKozCARsZM9LlxYc2H4YSvWn%2BSgA%3D&reserved=0>.
  *   but i added only the comments about the this commit
  *   just for information, this commit has following changes:
     *   renames ds_trafficsign to ds_tsr
     *   added graching data set
     *   ad new iterator ds_image which will handle both image and video input in util
     *   ds_kitti, ds_tsr, ds_dashcam use ds_image to handle both image and video input and provide slam input.
     *   change launch.vs.json for launch run_dashcam and run_tsr.
     *   run_ds_dashcam.bat and run_tsr.bat will launch with default graching dataset if not parameter is given
  *   let me know how to resolve this now.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23issuecomment-411321983&data=02%7C01%7C%7C600db7eccf4f4b1c6ca908d5fd05319c%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636693121099853168&sdata=m1193UlU4qSYqI6VKs7kJCREH2b9KKvwRjyC9IVzoRQ%3D&reserved=0>, or mute the thread<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKF8biFNHnj4Cm1-swM9Rkjr1qfT_2Ceks5uOprsgaJpZM4TJpfe&data=02%7C01%7C%7C600db7eccf4f4b1c6ca908d5fd05319c%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636693121099853168&sdata=uAVg%2BZXYczi%2Bg%2BXptvE1TPn3OBlg1rEKayS%2FbSbIffA%3D&reserved=0>.

apattnaik0721013 commented:
@paul-michalik 
- create a temporary_branch [Refactor-ORB_SLAM2-front-end-to-accept-arbitary-data-sources-and-data-sinks-temporary](https://github.com/paul-michalik/ORB_SLAM2/tree/Refactor-ORB_SLAM2-front-end-to-accept-arbitary-data-sources-and-data-sinks-temporary) from my latest changes.
- reset to your previous commit 6c202eb

shall i add the following changes:
- add Garching_LoopClosure-5-images dataset to data folder
- support boost-parameters in data source.

paul-michalik commented:
OK, I am giving up, you win. You obviously do not read what I write... So let me push the changes to this branch. You can add `Garching_LoopClosure-5-images` to the `feature/temp/Refactor-ORB_SLAM2-front-end-to-accept-arbitary-data-sources-and-data-sinks-temporary` branch. Please go and create a new pull request for the temporary branch and we are going to merge it onto this branch. But: **please do not merge anything until I tell you to do so. You will need to rewrite a lot of code and add unit tests for each single data source**

apattnaik0721013 commented:
@paul-michalik 
summarizing the list of task that i am going to do in `feature/temp/Refactor-ORB_SLAM2-front-end-to-accept-arbitary-data-sources-and-data-sinks-temporary`:
- add Garching_LoopClosure-5-images
- change the structural representation of the branch `feature/temp/Refactor-ORB_SLAM2-front-end-to-accept-arbitary-data-sources-and-data-sinks-temporary` as you did in this current branch:
  - move all the `data source` to `ext_ds`
  - all `data source` unit tests to `ext_tests`
- retaining the ds_image support in `feature/temp/Refactor-ORB_SLAM2-front-end-to-accept-arbitary-data-sources-and-data-sinks-temporary` which makes data source supporting both image and video inputs.

paul-michalik commented:
- Follow the implementation patterns for arguments and data sources from ds_tsr_args and ds_tsr
- do not modify these two in your branch, this is still work in progress
- apply zero warnings policy! No warnings must be used from the newly written code. CMake is set to W3 for MSVC
- write unit tests for data sources and arguments as demonstrated. With MSVC precompilation and ergo fast design cycles are supported
- expect a lot of redesign cycles since the submitted code is far from good

Sent from my Windows 10 device

________________________________
From: ansuman pattnaik <notifications@github.com>
Sent: Friday, August 10, 2018 10:12:29 AM
To: paul-michalik/ORB_SLAM2
Cc: Paul Michalik; Mention
Subject: Re: [paul-michalik/ORB_SLAM2] Feature/refactor orb slam2 front end to accept arbitary data sources and data sinks (#68)


@paul-michalik<https://nam01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik&data=02%7C01%7C%7Ca25a01afadb749971d5e08d5fe990458%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636694855506991844&sdata=Rkgi7K6a2Rs04Pzy6J2w0p6yhY8MfiIUoS%2FbOd%2BaLvo%3D&reserved=0>
summarizing the list of task that i am going to do in feature/temp/Refactor-ORB_SLAM2-front-end-to-accept-arbitary-data-sources-and-data-sinks-temporary:

  *   add Garching_LoopClosure-5-images
  *   change the structural representation of the branch feature/temp/Refactor-ORB_SLAM2-front-end-to-accept-arbitary-data-sources-and-data-sinks-temporary as you did in this current branch:
     *   move all the data source to ext_ds
     *   all data source unit tests to ext_tests
  *   retaining the ds_image support in feature/temp/Refactor-ORB_SLAM2-front-end-to-accept-arbitary-data-sources-and-data-sinks-temporary which makes data source supporting both image and video inputs.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpaul-michalik%2FORB_SLAM2%2Fpull%2F68%23issuecomment-412010376&data=02%7C01%7C%7Ca25a01afadb749971d5e08d5fe990458%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636694855507148096&sdata=oQaMeqyr%2Fr2yo%2B0qevnrIMNe%2BKX%2B7HE0TFE3%2BNPLH24%3D&reserved=0>, or mute the thread<https://nam01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKF8bsvy-b_Lc9wh1bO3Ti1Ot-Un4orwks5uPUBtgaJpZM4TJpfe&data=02%7C01%7C%7Ca25a01afadb749971d5e08d5fe990458%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636694855507148096&sdata=oAQdtmtYZGx8CRhfZx9d187idbXnjaP8V30qDPqsF%2BQ%3D&reserved=0>.

apattnaik0721013 commented:
Yes paul, we don't need separate tsr data.

paul-michalik committed:
Add more test and investigate short and long CLI conflict

paul-michalik committed:
Add exploration tests for Boost program options.

paul-michalik committed:
(1)  Track certain types of files via git-lfs (2) Add test data for ds_tsr (3) Add copying of test data to binary folder in order to support automatic testing

paul-michalik committed:
Add more tests and fix test execution parameters
